cmd/provider.go

```diff
@@ -23,56 +19,46 @@ var providerCmd = Cmd{
 func runProvider(args ...string) {
 	fs := flag.NewFlagSet("provider", flag.ExitOnError)
 	addrFlag := fs.String("addr", "", "address to listen at (host:port)")
+	localFlag := fs.String("local", "", "address to listen at for local provider api (host:port)")
 	fs.Parse(args)
 
 	homedir, err := findHomeDir()
 	if err != nil {
 		log.Fatal(err)
 	}
 
-	p, err := provider.LoadFromDisk(path.Join(homedir, "provider"))
+	logfile, err := os.OpenFile(path.Join(homedir, "provider", "provider.log"), os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
 	if err != nil {
-		log.Fatal(err)
+		log.Fatalf("Cannot create log file: %s", err)
 	}
+	defer logfile.Close()
+	logger := log.New(logfile, "", log.LstdFlags)
 
-	addr := *addrFlag
-	if len(addr) == 0 {
-		addr = p.Config.ApiAddr
+	p, err := provider.InitProvider(homedir)
+	if err != nil {
+		log.Fatalf("Failed to init provider: %s", err)
 	}
 
-	// Register with metadata service.
-	if !p.Config.IsRegistered {
-		pubKeyBytes, err := ioutil.ReadFile(p.Config.PublicKeyFile)
-		if err != nil {
-			log.Fatal("Could not read public key file. Error: ", err)
-		}
-		info := core.ProviderInfo{
-			PublicKey:  string(pubKeyBytes),
-			Addr:       addr,
-			SpaceAvail: 1 << 32,
-		}
-		metaService := metaserver.NewClient(p.Config.MetaAddr, &http.Client{})
-		config, err := metaService.RegisterProvider(&info)
-		if err != nil {
-			log.Fatalf("Unable to register with metaservice. Error: %s", err)
-		}
-		p.Config.ProviderID = config.ID
-		p.Config.IsRegistered = true
-		err = util.SaveJson(path.Join(homedir, "provider", "config.json"), p.Config)
-		if err != nil {
-			log.Fatalf("Unable to update config after registering with metaserver. Error: %s", err)
-		}
+	addr := *addrFlag
+	localAddr := *localFlag
+	// only run local facing server (statistics for ui) when localAddr is set
+	if len(localAddr) != 0 {
+		go func() {
+			// localAddr = "localhost:29876"
+			localServer := provider.NewLocalServer(p, logger)
+			log.Println("starting local provider server at", localAddr)
+			log.Fatal(http.ListenAndServe(localAddr, localServer))
+		}()
 	}
 
-	logfile, err := os.OpenFile(path.Join(homedir, "provider", "provider.log"), os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
-	if err != nil {
-		log.Fatalf("Cannot create log file: %s", err)
+	if len(addr) == 0 {
+		addr = p.Config.ApiAddr
+	} else {
+		// change provider configuration to use new public addr and update w/meta
+		p.Config.ApiAddr = addr
```

Comment:
I'm not sure we should permanently change the address if the -addr flag is given.I would probably lean towards either 1) removing the -addr flag altogether or 2) updating the metaserver with the new address but then going back to using the configured address if run again.

Reply:
At this point I think it is more of a convenience flag, so it could easily be removed. 

I think that if a provider updates their info on the metaserver it makes sense to persist with that address.

Reply:
I agree with Zak on this. It would be strange to me if I updated my information with the metaserver but the local application didn't persist that.